### PR TITLE
PSG-1875: Add genotyping results to sars-cov-2 pipeline output

### PR DESCRIPTION
**Pull request by pdp10**
* add genotyping results to pipeline results.csv. The new column is called `GENOTYPING` and is a dictionary.
* the resultfiles.json includes the 3 ncov typing output files
* the pipeline also outputs a results.json, which is like the results.csv but in json format.

**Example of output:**
Illumina:
```
root@sars-cov-2-pipeline-minikube-6cf7cbfc9c-zvwxg:/app/psga# cat /data/output/pipeline_ci_illumina/results.csv 
SAMPLE_ID,STATUS,CONTAMINATED_READS,PRIMER_DETECTED,PRIMER_NUMREADS,PRIMER_UNIQUE_NUMREADS,PRIMER_COVERAGE,PRIMER_INPUT,PCT_N_BASES,PCT_COVERED_BASES,LONGEST_NO_N_RUN,NUM_ALIGNED_READS,QC_PASS,GENOTYPING,LINEAGE,CONFLICT,AMBIGUITY_SCORE,SCORPIO_CALL,SCORPIO_SUPPORT,SCORPIO_CONFLICT,SCORPIO_NOTES,VERSION,PANGOLIN_VERSION,SCORPIO_VERSION,CONSTELLATION_VERSION,IS_DESIGNATED,QC_STATUS,QC_NOTES,NOTE
9729bce7-f0a9-4617-b6e0-6145307741d1,Completed,644,ARTIC_V4-1,453,66,0.6407766990291263,unknown,1.44,98.55,21625,265859,True,"{'D614G': {'num_matching_vars': '1', 'num_missing_vars': '0', 'type_coverage': '1.0', 'found_vars': 'S.D614G', 'missing_vars': '', 'additional_vars': 'ORF1ab.D691N;ORF1ab.Syn.924F;ORF1ab.A1306S;ORF1ab.N1555D;ORF1ab.P2046L;ORF1ab.P2287S;ORF1ab.Syn.2907D;ORF1ab.V2930L;ORF1ab.T3255I;ORF1ab.Syn.3585L;ORF1ab.T3646A;ORF1ab.Syn.3689V;ORF1ab.R3993H;ORF1ab.Syn.4715L;ORF1ab.Syn.5062A;ORF1ab.H5401Y;ORF1ab.R5689I;ORF1ab.Syn.6319L;ORF1ab.S6512I;S.T19R;S.EFR156G;S.P681R;S.D950N;ORF3a.S26L;M.I82T;ORF7a.V82A;ORF7a.T120I;ORF7b.T40I;ORF8.LDF118L;N.D63G;N.R203M;N.G215C;N.D377Y'}}",AY.6,0.0,,Delta (B.1.617.2-like),0.77,0.08,scorpio call: Alt alleles 10; Ref alleles 1; Amb alleles 2; Oth alleles 0,PUSHER-v1.16,4.1.3,0.3.17,v0.1.10,False,pass,Ambiguous_content:0.03,Usher placements: AY.6(2/2)

root@sars-cov-2-pipeline-minikube-6cf7cbfc9c-zvwxg:/app/psga# cat /data/output/pipeline_ci_illumina/results.json
{
    "9729bce7-f0a9-4617-b6e0-6145307741d1": {
        "AMBIGUITY_SCORE": "",
        "CONFLICT": "0.0",
        "CONSTELLATION_VERSION": "v0.1.10",
        "CONTAMINATED_READS": "644",
        "GENOTYPING": {
            "D614G": {
                "additional_vars": "ORF1ab.D691N;ORF1ab.Syn.924F;ORF1ab.A1306S;ORF1ab.N1555D;ORF1ab.P2046L;ORF1ab.P2287S;ORF1ab.Syn.2907D;ORF1ab.V2930L;ORF1ab.T3255I;ORF1ab.Syn.3585L;ORF1ab.T3646A;ORF1ab.Syn.3689V;ORF1ab.R3993H;ORF1ab.Syn.4715L;ORF1ab.Syn.5062A;ORF1ab.H5401Y;ORF1ab.R5689I;ORF1ab.Syn.6319L;ORF1ab.S6512I;S.T19R;S.EFR156G;S.P681R;S.D950N;ORF3a.S26L;M.I82T;ORF7a.V82A;ORF7a.T120I;ORF7b.T40I;ORF8.LDF118L;N.D63G;N.R203M;N.G215C;N.D377Y",
                "found_vars": "S.D614G",
                "missing_vars": "",
                "num_matching_vars": "1",
                "num_missing_vars": "0",
                "type_coverage": "1.0"
            }
        },
        "IS_DESIGNATED": "False",
        "LINEAGE": "AY.6",
        "LONGEST_NO_N_RUN": "21625",
        "NOTE": "Usher placements: AY.6(2/2)",
        "NUM_ALIGNED_READS": "265859",
        "PANGOLIN_VERSION": "4.1.3",
        "PCT_COVERED_BASES": "98.55",
        "PCT_N_BASES": "1.44",
        "PRIMER_COVERAGE": "0.6407766990291263",
        "PRIMER_DETECTED": "ARTIC_V4-1",
        "PRIMER_INPUT": "unknown",
        "PRIMER_NUMREADS": "453",
        "PRIMER_UNIQUE_NUMREADS": "66",
        "QC_NOTES": "Ambiguous_content:0.03",
        "QC_PASS": "True",
        "QC_STATUS": "pass",
        "SCORPIO_CALL": "Delta (B.1.617.2-like)",
        "SCORPIO_CONFLICT": "0.08",
        "SCORPIO_NOTES": "scorpio call: Alt alleles 10; Ref alleles 1; Amb alleles 2; Oth alleles 0",
        "SCORPIO_SUPPORT": "0.77",
        "SCORPIO_VERSION": "0.3.17",
        "STATUS": "Completed",
        "VERSION": "PUSHER-v1.16"
    }
}
```

ONT:
```
root@sars-cov-2-pipeline-minikube-6cf7cbfc9c-zvwxg:/app/psga# cat /data/output/pipeline_ci_ont/results.csv 
SAMPLE_ID,STATUS,CONTAMINATED_READS,PRIMER_DETECTED,PRIMER_NUMREADS,PRIMER_UNIQUE_NUMREADS,PRIMER_COVERAGE,PRIMER_INPUT,PCT_N_BASES,PCT_COVERED_BASES,LONGEST_NO_N_RUN,NUM_ALIGNED_READS,QC_PASS,GENOTYPING,LINEAGE,CONFLICT,AMBIGUITY_SCORE,SCORPIO_CALL,SCORPIO_SUPPORT,SCORPIO_CONFLICT,SCORPIO_NOTES,VERSION,PANGOLIN_VERSION,SCORPIO_VERSION,CONSTELLATION_VERSION,IS_DESIGNATED,QC_STATUS,QC_NOTES,NOTE
d3e958b5-0c56-48d6-807b-6a89f61408ce,Completed,4,none,0,0,0,unknown,23.08,76.55,2201,19057,True,"{'D614G': {'num_matching_vars': '1', 'num_missing_vars': '0', 'type_coverage': '1.0', 'found_vars': 'S.D614G', 'missing_vars': '', 'additional_vars': 'ORF1ab.Syn.216S;ORF1ab.Syn.924F;ORF1ab.Frameshift.958KHLWNLVPLLLLFNLKKSKKKIG*;S.A570D;S.S982A;S.D1118H;ORF8.Q27*;ORF8.R52I;ORF8.Y73C'}}",B.1.1.7,0.0,,,,,,PUSHER-v1.16,4.1.3,0.3.17,v0.1.10,False,pass,Ambiguous_content:0.24,Usher placements: B.1.1.7(10/10)
2cf78ed4-edae-4943-ab04-5d7a1e4cbdcb,Completed,7,none,0,0,0,unknown,12.35,87.71,3925,33005,True,"{'VOC-202012/01': {'num_matching_vars': '12', 'num_missing_vars': '5', 'type_coverage': '0.71', 'found_vars': 'S.VY143V;S.A570D;S.S982A;S.D1118H;ORF1ab.T1001I;ORF1ab.A1708D;ORF1ab.I2230T;ORF1ab.LSGF3674L;ORF8.Q27*;ORF8.R52I;ORF8.Y73C;N.D3L', 'missing_vars': 'S.IHV68I;S.N501Y;S.P681H;S.T716I;N.S235F', 'additional_vars': 'ORF1ab.Syn.216S;ORF1ab.L730F;ORF1ab.Syn.924F;ORF1ab.Syn.1907F;ORF1ab.V2586G;ORF1ab.Syn.2587A;ORF1ab.I3737V;ORF1ab.Syn.4715L;ORF1ab.P4804L;ORF1ab.T5005I;ORF1ab.L5304P;S.D614G;ORF6.Syn.4L'}, 'D614G': {'num_matching_vars': '1', 'num_missing_vars': '0', 'type_coverage': '1.0', 'found_vars': 'S.D614G', 'missing_vars': '', 'additional_vars': 'ORF1ab.Syn.216S;ORF1ab.L730F;ORF1ab.Syn.924F;ORF1ab.T1001I;ORF1ab.A1708D;ORF1ab.Syn.1907F;ORF1ab.I2230T;ORF1ab.V2586G;ORF1ab.Syn.2587A;ORF1ab.LSGF3674L;ORF1ab.I3737V;ORF1ab.Syn.4715L;ORF1ab.P4804L;ORF1ab.T5005I;ORF1ab.L5304P;S.VY143V;S.A570D;S.S982A;S.D1118H;ORF6.Syn.4L;ORF8.Q27*;ORF8.R52I;ORF8.Y73C;N.D3L'}}",B.1.1.7,0.0,,Alpha (B.1.1.7-like),0.74,0.04,scorpio call: Alt alleles 17; Ref alleles 1; Amb alleles 5; Oth alleles 0,PUSHER-v1.16,4.1.3,0.3.17,v0.1.10,False,pass,Ambiguous_content:0.14,Usher placements: B.1.1.7(1/1)

root@sars-cov-2-pipeline-minikube-6cf7cbfc9c-zvwxg:/app/psga# cat /data/output/pipeline_ci_ont/results.json
{
    "2cf78ed4-edae-4943-ab04-5d7a1e4cbdcb": {
        "AMBIGUITY_SCORE": "",
        "CONFLICT": "0.0",
        "CONSTELLATION_VERSION": "v0.1.10",
        "CONTAMINATED_READS": "7",
        "GENOTYPING": {
            "D614G": {
                "additional_vars": "ORF1ab.Syn.216S;ORF1ab.L730F;ORF1ab.Syn.924F;ORF1ab.T1001I;ORF1ab.A1708D;ORF1ab.Syn.1907F;ORF1ab.I2230T;ORF1ab.V2586G;ORF1ab.Syn.2587A;ORF1ab.LSGF3674L;ORF1ab.I3737V;ORF1ab.Syn.4715L;ORF1ab.P4804L;ORF1ab.T5005I;ORF1ab.L5304P;S.VY143V;S.A570D;S.S982A;S.D1118H;ORF6.Syn.4L;ORF8.Q27*;ORF8.R52I;ORF8.Y73C;N.D3L",
                "found_vars": "S.D614G",
                "missing_vars": "",
                "num_matching_vars": "1",
                "num_missing_vars": "0",
                "type_coverage": "1.0"
            },
            "VOC-202012/01": {
                "additional_vars": "ORF1ab.Syn.216S;ORF1ab.L730F;ORF1ab.Syn.924F;ORF1ab.Syn.1907F;ORF1ab.V2586G;ORF1ab.Syn.2587A;ORF1ab.I3737V;ORF1ab.Syn.4715L;ORF1ab.P4804L;ORF1ab.T5005I;ORF1ab.L5304P;S.D614G;ORF6.Syn.4L",
                "found_vars": "S.VY143V;S.A570D;S.S982A;S.D1118H;ORF1ab.T1001I;ORF1ab.A1708D;ORF1ab.I2230T;ORF1ab.LSGF3674L;ORF8.Q27*;ORF8.R52I;ORF8.Y73C;N.D3L",
                "missing_vars": "S.IHV68I;S.N501Y;S.P681H;S.T716I;N.S235F",
                "num_matching_vars": "12",
                "num_missing_vars": "5",
                "type_coverage": "0.71"
            }
        },
        "IS_DESIGNATED": "False",
        "LINEAGE": "B.1.1.7",
        "LONGEST_NO_N_RUN": "3925",
        "NOTE": "Usher placements: B.1.1.7(1/1)",
        "NUM_ALIGNED_READS": "33005",
        "PANGOLIN_VERSION": "4.1.3",
        "PCT_COVERED_BASES": "87.71",
        "PCT_N_BASES": "12.35",
        "PRIMER_COVERAGE": "0",
        "PRIMER_DETECTED": "none",
        "PRIMER_INPUT": "unknown",
        "PRIMER_NUMREADS": "0",
        "PRIMER_UNIQUE_NUMREADS": "0",
        "QC_NOTES": "Ambiguous_content:0.14",
        "QC_PASS": "True",
        "QC_STATUS": "pass",
        "SCORPIO_CALL": "Alpha (B.1.1.7-like)",
        "SCORPIO_CONFLICT": "0.04",
        "SCORPIO_NOTES": "scorpio call: Alt alleles 17; Ref alleles 1; Amb alleles 5; Oth alleles 0",
        "SCORPIO_SUPPORT": "0.74",
        "SCORPIO_VERSION": "0.3.17",
        "STATUS": "Completed",
        "VERSION": "PUSHER-v1.16"
    },
    "d3e958b5-0c56-48d6-807b-6a89f61408ce": {
        "AMBIGUITY_SCORE": "",
        "CONFLICT": "0.0",
        "CONSTELLATION_VERSION": "v0.1.10",
        "CONTAMINATED_READS": "4",
        "GENOTYPING": {
            "D614G": {
                "additional_vars": "ORF1ab.Syn.216S;ORF1ab.Syn.924F;ORF1ab.Frameshift.958KHLWNLVPLLLLFNLKKSKKKIG*;S.A570D;S.S982A;S.D1118H;ORF8.Q27*;ORF8.R52I;ORF8.Y73C",
                "found_vars": "S.D614G",
                "missing_vars": "",
                "num_matching_vars": "1",
                "num_missing_vars": "0",
                "type_coverage": "1.0"
            }
        },
        "IS_DESIGNATED": "False",
        "LINEAGE": "B.1.1.7",
        "LONGEST_NO_N_RUN": "2201",
        "NOTE": "Usher placements: B.1.1.7(10/10)",
        "NUM_ALIGNED_READS": "19057",
        "PANGOLIN_VERSION": "4.1.3",
        "PCT_COVERED_BASES": "76.55",
        "PCT_N_BASES": "23.08",
        "PRIMER_COVERAGE": "0",
        "PRIMER_DETECTED": "none",
        "PRIMER_INPUT": "unknown",
        "PRIMER_NUMREADS": "0",
        "PRIMER_UNIQUE_NUMREADS": "0",
        "QC_NOTES": "Ambiguous_content:0.24",
        "QC_PASS": "True",
        "QC_STATUS": "pass",
        "SCORPIO_CALL": "",
        "SCORPIO_CONFLICT": "",
        "SCORPIO_NOTES": "",
        "SCORPIO_SUPPORT": "",
        "SCORPIO_VERSION": "0.3.17",
        "STATUS": "Completed",
        "VERSION": "PUSHER-v1.16"
    }
}
```

FASTA input:
```
root@sars-cov-2-pipeline-minikube-6cf7cbfc9c-zvwxg:/app/psga# cat /data/output/pipeline_ci_unknown/results.csv
SAMPLE_ID,STATUS,CONTAMINATED_READS,PRIMER_NUMREADS,PRIMER_DETECTED,PRIMER_UNIQUE_NUMREADS,PRIMER_INPUT,PRIMER_COVERAGE,PCT_COVERED_BASES,LONGEST_NO_N_RUN,QC_PASS,NUM_ALIGNED_READS,PCT_N_BASES,GENOTYPING,LINEAGE,CONFLICT,AMBIGUITY_SCORE,SCORPIO_CALL,SCORPIO_SUPPORT,SCORPIO_CONFLICT,SCORPIO_NOTES,VERSION,PANGOLIN_VERSION,SCORPIO_VERSION,CONSTELLATION_VERSION,IS_DESIGNATED,QC_STATUS,QC_NOTES,NOTE
19fc4e8f-982c-4679-a37a-0501f3ccb7cc,Completed,,,,,,,,,,,,,AY.6,0.0,,Delta (B.1.617.2-like),0.77,0.0,scorpio call: Alt alleles 10; Ref alleles 0; Amb alleles 3; Oth alleles 0,PUSHER-v1.16,4.1.3,0.3.17,v0.1.10,False,pass,Ambiguous_content:0.03,Usher placements: AY.6(2/2)
1a55b645-5cce-400d-b668-e281a95b4c72,Completed,,,,,,,,,,,,,AY.98,0.0,,Delta (B.1.617.2-like),1.0,0.0,scorpio call: Alt alleles 13; Ref alleles 0; Amb alleles 0; Oth alleles 0,PUSHER-v1.16,4.1.3,0.3.17,v0.1.10,False,pass,Ambiguous_content:0.04,Usher placements: AY.98(1/1)

root@sars-cov-2-pipeline-minikube-6cf7cbfc9c-zvwxg:/app/psga# cat /data/output/pipeline_ci_unknown/results.json
{
    "19fc4e8f-982c-4679-a37a-0501f3ccb7cc": {
        "AMBIGUITY_SCORE": "",
        "CONFLICT": "0.0",
        "CONSTELLATION_VERSION": "v0.1.10",
        "CONTAMINATED_READS": "",
        "GENOTYPING": "",
        "IS_DESIGNATED": "False",
        "LINEAGE": "AY.6",
        "LONGEST_NO_N_RUN": "",
        "NOTE": "Usher placements: AY.6(2/2)",
        "NUM_ALIGNED_READS": "",
        "PANGOLIN_VERSION": "4.1.3",
        "PCT_COVERED_BASES": "",
        "PCT_N_BASES": "",
        "PRIMER_COVERAGE": "",
        "PRIMER_DETECTED": "",
        "PRIMER_INPUT": "",
        "PRIMER_NUMREADS": "",
        "PRIMER_UNIQUE_NUMREADS": "",
        "QC_NOTES": "Ambiguous_content:0.03",
        "QC_PASS": "",
        "QC_STATUS": "pass",
        "SCORPIO_CALL": "Delta (B.1.617.2-like)",
        "SCORPIO_CONFLICT": "0.0",
        "SCORPIO_NOTES": "scorpio call: Alt alleles 10; Ref alleles 0; Amb alleles 3; Oth alleles 0",
        "SCORPIO_SUPPORT": "0.77",
        "SCORPIO_VERSION": "0.3.17",
        "STATUS": "Completed",
        "VERSION": "PUSHER-v1.16"
    },
    "1a55b645-5cce-400d-b668-e281a95b4c72": {
        "AMBIGUITY_SCORE": "",
        "CONFLICT": "0.0",
        "CONSTELLATION_VERSION": "v0.1.10",
        "CONTAMINATED_READS": "",
        "GENOTYPING": "",
        "IS_DESIGNATED": "False",
        "LINEAGE": "AY.98",
        "LONGEST_NO_N_RUN": "",
        "NOTE": "Usher placements: AY.98(1/1)",
        "NUM_ALIGNED_READS": "",
        "PANGOLIN_VERSION": "4.1.3",
        "PCT_COVERED_BASES": "",
        "PCT_N_BASES": "",
        "PRIMER_COVERAGE": "",
        "PRIMER_DETECTED": "",
        "PRIMER_INPUT": "",
        "PRIMER_NUMREADS": "",
        "PRIMER_UNIQUE_NUMREADS": "",
        "QC_NOTES": "Ambiguous_content:0.04",
        "QC_PASS": "",
        "QC_STATUS": "pass",
        "SCORPIO_CALL": "Delta (B.1.617.2-like)",
        "SCORPIO_CONFLICT": "0.0",
        "SCORPIO_NOTES": "scorpio call: Alt alleles 13; Ref alleles 0; Amb alleles 0; Oth alleles 0",
        "SCORPIO_SUPPORT": "1.0",
        "SCORPIO_VERSION": "0.3.17",
        "STATUS": "Completed",
        "VERSION": "PUSHER-v1.16"
    }
}
```

### CODE REVIEW

File: psga/config/psga-schema-sars-cov-2.yaml [hunk 1]
```diff
@@ -322,6 +322,10 @@ output_fields:
     field_type: string
     required: true
 
+  - field_name: GENOTYPING
+    field_type: dictionary
+    required: true
+
```
Comment (pdp10) [inline psga/config/psga-schema-sars-cov-2.yaml:328]: I guessed this field_type.

Comment (cramshaw) [reply psga/config/psga-schema-sars-cov-2.yaml:328]: Need to check with @Aiky30 if this will break anything

File: scripts/util/convert.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+from pathlib import Path
+import csv
+import ast
+import json
+
+
+def csv_to_json(csv_path: Path, json_path: Path, sample_id: str):
+    """
+    Convert a CSV file to a JSON file using sample_id as key in the JSON
```
Comment (cramshaw) [inline scripts/util/convert.py:9]: If it makes it easier, you can just make the JSON an array of rows - almost exactly the output from csv.DictReader? I assume the nested dict in the csv causes an issue there?

File: scripts/sars_cov_2/generate_pipeline_results_files.py [hunk 3]
```diff
@@ -623,25 +657,30 @@ def generate_pipeline_results_files(
...
-        df_ncov,
+        df_ncov_qc,
+        df_ncov_typing,
         df_pangolin,
         qc_unrelated_failing_samples,
-        output_csv_file,
+        output_results_csv_file,
     )
 
+    csv_to_json(output_results_csv_file, output_results_json_file, SAMPLE_ID)
```
Comment (cramshaw) [inline scripts/sars_cov_2/generate_pipeline_results_files.py:682]: ideally this would be called by every pipeline

Comment (pdp10) [reply scripts/sars_cov_2/generate_pipeline_results_files.py:682]: That's correct. The code is in the utility dir.

File: psga/config/psga-schema-sars-cov-2.yaml (hunk 1 above)
Comment (Aiky30) [reply psga/config/psga-schema-sars-cov-2.yaml:328]: We only check for type on the date, it's not used anywhere else. There is a map to handle this in the future: https://github.com/Congenica/psga-backend/blob/5201b2c607ebdcd5514f572482553a06d5d4a6ef/data_management/metadata_validator.py#L42

Comment (Aiky30) [reply psga/config/psga-schema-sars-cov-2.yaml:328]: We will need to port this change into the PSGA-Backend, and then bring the config from the backend back into this project. They are very far apart atm where the backend has moved on.

Comment (cramshaw) [reply psga/config/psga-schema-sars-cov-2.yaml:328]: Devops are in progress on getting pipelines to install using their actual config. as this is a change to results (which isn't used currently) I think we can wait for that, but good call out.
From the sound of it this should be safe to go with for now though

**Review by cramshaw — APPROVED** [review]

**Pull request merged**